src/unity/toolkits/drawing_recognition/data_preparation.cpp

```diff
@@ -0,0 +1,262 @@
+/* Copyright © 2017 Apple Inc. All rights reserved.
+ *
+ * Use of this source code is governed by a BSD-3-clause license that can
+ * be found in the LICENSE.txt file or at https://opensource.org/licenses/BSD-3-Clause
+ */
+#include <time.h>
+#include <limits.h>
+#include <math.h>
+#include <iomanip>
+#include <logger/assertions.hpp>
+#include "data_preparation.hpp"
+#include "random/random.hpp"
+#include "util/sys_util.hpp"
+
+namespace turi {
+namespace drawing_recognition {
+
+struct Point {
+    float x;
+    float y;
+};
+
+struct Line {
+    // ax + by + c = 0
+    float slope;
+    float a;
+    float b;
+    float c;
+};
+
+void make_point(Point *P, const flex_dict &point_dict) {
+    flex_string x ("x");
+    flex_string y ("y");
+    for (auto key_value_pair: point_dict) {
+        const flex_string &key = key_value_pair.first.get<flex_string>();
+        const flex_float &value = key_value_pair.second.get<flex_float>();
+        if (key.compare(x) == 0) {
+            P->x = value;
+        } else {
+            P->y = value;
+        }
+    }
+}
+
+float line_to_point_distance(Line *L, const flex_dict &point) {
+    Point P; make_point(&P, point);
+    int x = (int)(P.x);
+    int y = (int)(P.y);
+    float a = L->a; float b = L->b; float c = L->c;
+    float distance = (float)(fabs(a*x+b*y+c))/sqrt(a*a+b*b);
+    return (int)distance;
+}
+
+void initialize_line(Line *L, const flex_dict &p1, const flex_dict &p2) {
+    Point P1; make_point(&P1, p1);
+    Point P2; make_point(&P2, p2);
+    float start_x = P1.x;
+    float start_y = P1.y;
+    float end_x = P2.x;
+    float end_y = P2.y;
+    if (start_x == start_y) {
+        L->slope = INT_MAX;
+    } else {
+        L->slope = ((float)(end_y - start_y))/((float)(end_x - start_x));
+    }
+    L->a = L->slope;
+    L->b = -1;
+    L->c = start_y - L->slope * start_x;
+}
+
+template<typename T>
+std::vector<T> slice(std::vector<T> const &v, int m, int n) {
+    auto first = v.cbegin() + m;
+    auto last = v.cbegin() + n + 1;
+    std::vector<T> vec(first, last);
+    return vec;
+}
+
+flex_list ramer_douglas_peucker(flex_list stroke, float epsilon) {
+    float dmax = 0;
```

Comment:
A comment with a link to wikipedia would be cool: https://en.wikipedia.org/wiki/Ramer–Douglas–Peucker_algorithm